### ISSUE TITLE
A(T) is not a generic class, it's a generic module

### ISSUE THREAD

**Issue by Ragmaanir** [crystal-lang/crystal#1900]
Hi,

not sure if there is an issue for this yet (couldn't find one):

```crystal
module A(T)
end

arr = [] of A(Int32)
```
Error:
```
A(T) is not a generic class, it's a generic module
```

Is this a bug or does it have a specific reason? If the latter, are there alternative ways to implement this?

Thanks!

**Comment by Perelandric** [comment]
A module can't be instantiated. Is there a reason you wouldn't use a class?

**Comment by omninonsense** [comment]
You can't instantiate a Module, you would use that as:

~~~crystal
module M(T)
end

class C(T)
  include M(T)
end

x = [] of C(Int32)
~~~

What are you trying to achieve, though?

**Comment by asterite** [comment]
This is kind of a bug or enhancement request, you can't have an array of generic modules. But in the next version of the language this will probably be supported.

**Comment by Ragmaanir** [comment]
@omninonsense:
I don't want to instantiate the module. I just want an array of objects that mix in that module, so they have the same interface. This is important since i don't want users of my library to have to inherit from a class.

@asterite:
Ok, that sounds good.

**Comment by kirbyfan64** [comment]
@Ragmaanir Ah, so you want structural typing?

**Comment by stugol** [comment]
+1 for this.

It's not a feature, it's a bug. It should have the bug label. The inability to have an array of `Enumerable(T)`, or a proc that takes an `Enumerable(T)` argument is a pretty serious oversight!

I needed this yesterday.

**Comment by mrkaspa** [comment]
Can a module act as interface to define methods to be implemented like GoLang?

**Comment by asterite** [comment]
@mrkaspa Yes, you can do that right now. You can't with generic modules, but that's only a temporary limitation, you'll be able to do that in the future.

**Comment by mrkaspa** [comment]
Cool, I didn't find that in the docs I thought It wasn't supported

**Comment by asterite** [comment]
Closed in favor of #2733
